core/src/main/java/com/linecorp/armeria/common/HttpHeaders.java

```diff
@@ -13,227 +13,156 @@
  * License for the specific language governing permissions and limitations
  * under the License.
  */
-
 package com.linecorp.armeria.common;
 
-import static java.util.Objects.requireNonNull;
-
-import java.util.Iterator;
-import java.util.Map.Entry;
-
-import javax.annotation.Nullable;
+import java.util.function.Consumer;
 
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 
-import io.netty.handler.codec.Headers;
-import io.netty.util.AsciiString;
-
 /**
- * HTTP/2 headers.
+ * Immutable HTTP/2 headers.
+ *
+ * @see RequestHeaders
+ * @see ResponseHeaders
  */
 @JsonSerialize(using = HttpHeadersJsonSerializer.class)
 @JsonDeserialize(using = HttpHeadersJsonDeserializer.class)
-public interface HttpHeaders extends HttpObject, Headers<AsciiString, String, HttpHeaders> {
-
-    /**
-     * An immutable empty HTTP/2 headers.
-     */
-    HttpHeaders EMPTY_HEADERS = new DefaultHttpHeaders(false, 0).asImmutable();
+public interface HttpHeaders extends HttpObject, HttpHeaderGetters {
 
     /**
-     * Returns new empty HTTP headers.
+     * Creates a new empty builder.
      */
-    static HttpHeaders of() {
-        return new DefaultHttpHeaders();
+    static HttpHeadersBuilder builder() {
+        return new DefaultHttpHeadersBuilder();
     }
 
     /**
-     * Returns new HTTP request headers.
+     * An empty {@link HttpHeaders}.
+     *
+     * @deprecated Use {@link #of()}.
      */
-    static HttpHeaders of(HttpMethod method, String path) {
-        return new DefaultHttpHeaders().method(method).path(path);
-    }
+    @Deprecated
+    HttpHeaders EMPTY_HEADERS = of();
 
     /**
-     * Returns new HTTP response headers.
+     * Returns an empty singleton {@link HttpHeaders}.
      */
-    static HttpHeaders of(int statusCode) {
-        return of(HttpStatus.valueOf(statusCode));
+    static HttpHeaders of() {
+        return DefaultHttpHeaders.EMPTY;
     }
 
     /**
-     * Returns new HTTP response headers.
+     * Returns a new {@link HttpHeaders} with the specified header.
      */
-    static HttpHeaders of(HttpStatus status) {
-        return new DefaultHttpHeaders().status(status);
+    static HttpHeaders of(CharSequence name, String value) {
+        return builder().add(name, value).build();
     }
 
     /**
-     * Returns new HTTP headers with a single entry.
+     * Returns a new {@link HttpHeaders} with the specified header. The value is converted into
+     * a {@link String} via {@link HttpHeadersBase#addObject(CharSequence, Object)}.
      */
-    static HttpHeaders of(AsciiString name, String value) {
-        return new DefaultHttpHeaders().add(name, value);
+    static HttpHeaders of(CharSequence name, Object value) {
```

Comment:
What do you think about keeping the methods of type `AsciiString` and `@Deprecated` them?

Reply:
Would it be worth doing that given that a user will have to recompile?

Reply:
Ah you're right, while I wanted to avoid having to recompile, it means picking the right method requires a cast which isn't good.

`NoSuchMethodError` at runtime is the scariest thing that can happen when upgrading a dependency, but guess there isn't much we can do here.

On the bright side, I guess most common uses of this class use the status or method variants and these might not be used so much in the wild.